### PR TITLE
fix(sec): Correctly hide passwords in command

### LINKED ISSUE #7949: "Unable to hide passwords in command"
Hi,

In 19.10.0-rc.1, #7859 leads to the following :
'Unable to hide passwords in command'

More details here : https://github.com/centreon/centreon/pull/7859#issuecomment-539950386

I opened this to keep track of this bug.

Many thx 👍

Edit : issue still present in 19.10.0-rc.1.

### PR DESCRIPTION
**Pull request by UrBnW**
Hi,

This fixes the "Unable to hide passwords in command" message in service details page.

It fixes #7949.

We did some debug here with @prohand (thank you !), it ends [from this post](https://github.com/centreon/centreon/pull/7975#issuecomment-544279124) below.

## Description

If a command contains macros which have spaces in their values (which is rather common), this leads to the following message in the service details page :

Screen Shot 2019-10-17 at 08 20 13

This PR then solves this. Here's how :

For every password macro :
- A = number of times macro name is found in command_with_macros ;
- B = number of times macro values (*) are replaced in command_executed ;
- if A != B, then throw an error ;
- Return command_executed.

(*) we can have several different values if macro is redefined.

The only case where an error is thrown is then when password used is too simple.
For example _pass_, and the command looks like : `test.pl --password=$_HOSTPWD$`
Otherwise it will succeed.
Use strong passwords and you won't face this case :)

We also try as much as possible to hide passwords from Centreon Plugins options which may be used in non-password macros.

Finally, instead of the `***` pattern to hide the password, we now use `$pw`, which is really convenient when we copy/past the command into a shell.

## Type of change

- [x] Patch fixing an issue (non-breaking change)

## Target serie

- [ ] 19.04.x
- [x] 19.10.x
- [x] 20.04.x
- [x] 20.10.x (master)

<h2> How this pull request can be tested ? </h2>

Use a command with several macros, several password macros, and macros containing spaces.
Without this PR, you'll receive the "Unable to hide passwords in command" message.
With this PR, all passwords are correctly replaced.

Thank you 👍

Edit : issue still present in 19.10.12.

### CODE REVIEW

**Comment by prohand** [comment]
Hello,

Same issue for me.

Thank you

**Comment by CPbN** [comment]
This PR does not fully solve the issue though, I'm on it.

**Comment by CPbN** [comment]
OK, here's a working version.
How it works :

For every password macro :
- A = number of times macro name is used in command_with_macros ;
- B = number of times macro value is used in command_executed ;
- if A == B, then replace macro value in command_executed ;
- otherwise throw an error ;
- return command_executed.

The only case where an error is thrown is when password used is too simple, for example pass.
And the command looks like : `test.pl --password=$_HOSTPWD$`
Otherwise it will succeed.
Use strong passwords :)

Thx 👍

**Comment by prohand** [comment]
Thanks, I use a strong password but my password macro it's not hide in command line.
I use 19.10.1.

**Comment by CPbN** [comment]
Did you try to apply this patch ?

**Comment by prohand** [comment]
These patches:
https://github.com/centreon/centreon/commit/6a6920121f904b20113884e0272375d4573151cc
https://github.com/centreon/centreon/commit/d27945513e03890fdb123e272937df26807630a4

I have to do it manually? Or there is a command to do it automatically?

Thank you

**Comment by CPbN** [comment]
The easiest way :
```
cd /usr/share/centreon
wget https://github.com/centreon/centreon/pull/7975.diff
patch -p1 <7975.diff
```

**Comment by prohand** [comment]
Thank you.
I apply this patch but the password is show

**Comment by CPbN** [comment]
Is your password in a password-type macro like the following ?

Screen Shot 2019-10-18 at 09 05 08

**Comment by prohand** [comment]
Yes :  

image

**Comment by CPbN** [comment]
At which level is it ? Host ? Host template ? Service ? Service template ?

**Comment by CPbN** [comment]
And what does your command definition looks like please ?

**Comment by prohand** [comment]
image

**Comment by CPbN** [comment]
For now I don't understand why this occurs.
I have password-macros in host templates, at different inherited levels, and they are correctly hidden.
Looks weird 😕

**Comment by prohand** [comment]
Yes very weird. However, these passwords were entered before the patch

**Comment by CPbN** [comment]
You can revert the patch for now if you want.
Simply add `-R` into the _patch_ command above.

**Comment by prohand** [comment]
Ok it's done

**Comment by CPbN** [comment]
I just pushed a slightly different version, which is even simpler.
You can test it.

**Comment by prohand** [comment]
I repatch it : 

cd /usr/share/centreon
wget https://github.com/centreon/centreon/pull/7975.diff
patch -p1 <7975.diff

But no working again :(

**Comment by CPbN** [comment]
Password still shows ?
Try again (revert and re-apply), let's test a case-insensitive macro replacement...

**Comment by CPbN** [comment]
You then re-applied a third time, after my third commit, right ?

Just to be sure, reload your poller configuration, force a service check, wait for the check to proceed, and look at the service details.
Password still shows ?

**Comment by prohand** [comment]
Yes, i reload the conf and force service check. 
Same problem

**Comment by CPbN** [comment]
Could you show me a screenshot of command line in service details ?
_Of course hide sensitive data if you prefer..._

**Comment by prohand** [comment]
Yes : 
image

**Comment by CPbN** [comment]
New version, could you please test when you'll have a few minutes ?
Thank you 👍

**Comment by prohand** [comment]
Same problem :(

**Comment by CPbN** [comment]
!
Revert the current patch.
Then apply this debug one :
```
cd /usr/share/centreon
wget https://github.com/centreon/centreon/commit/f396b2e76ce6980c70b0eec75ded7ac40633bf4b.diff
patch -p1 <f396b2e76ce6980c70b0eec75ded7ac40633bf4b.diff
```
Then please do a screenshot of command line in service details.
It should contain raw command with macros, and executed command.
_Of course feel free to hide sensitive data._

Then revert the debug patch.

Thank you 👍

**Comment by prohand** [comment]
I just applied this patch
I have that in the service details : 
Image 8

**Comment by CPbN** [comment]
OK, interesting.
Strangely enough there are some difficulties to retrieve your command from the database.
You can re-apply the current patch.
You should now get a proper error message including your command name.
Is the command name returned the one you use for this service ?
Let's assume your command name is _mssql_, what do the following 2 SQL queries return ?
`select * from command where command_name = "mssql";`
`select * from command where command_name like "%mssql%";`
_On your server, with root user, `mysql centreon` to go for a MySQL prompt._

**Comment by prohand** [comment]
Here is the error message:

image

Yes, this is the name of the order used for the service

image

Thanks

**Comment by CPbN** [comment]
Really helpful debug, many thanks !
I just pushed a commit which should solve the issue 👍

**Comment by prohand** [comment]
Unfortunately I still have the problem :(

**Comment by CPbN** [comment]
Sounds like you did not apply the correct patch.
Please apply this PR, not the debug patch we tested before.

**Comment by prohand** [comment]
wget https://github.com/centreon/centreon/pull/7975.diff
patch -p1 <7975.diff

it's not correct ?

**Comment by CPbN** [comment]
It is.
But be sure to revert the previously applied patch before applying the new one (so that you revert to the original / non-modified file).
And be sure to remove the existing _7975.diff_ file before getting the new one, otherwise wget will add a _.1_ suffix to the new downloaded filename.
If you've messed-up your original file, you can do the following to restore it :
```
wget https://raw.githubusercontent.com/centreon/centreon/19.10.1/www/include/monitoring/objectDetails/common-func.php -O /usr/share/centreon/www/include/monitoring/objectDetails/common-func.php
```

**Comment by prohand** [comment]
It's OK :

image

Thanks ;)

**Comment by CPbN** [comment]
Good news, perfect, thank you for your help and feedback 👍

**Comment by o71** [comment]
Heelo,
Thank you very much, it solved my problem.

**Comment by carlosba1982** [comment]
Is this patch eventually hitting repos so I could just update and solve this behaviour? I prefer wait than patching..

If it helps, I also realised something: I'm using latest stable Centreon 19.10.1 (commit e77f8037b) + Centreon Engine 19.10.3, "Remote Servers" and macros with ofuscated passwords. When I check in Central web "Executed Command Line" I see the message "Unable to hide passwords in command" as talked here, but if go to the web interface of the "Remote Server" which is feeding Central, the executed command line for that same exact host/service is correctly shown and password is shown in plain text.  

I prefer executed command + password revealed than an empty field. It makes troubleshooting tasks easier and In my case only monitoring administrator are able to see the field executed command line.

**Comment by CPbN** [comment]
Patch is ready to land, waiting for review :wink:

**Comment by CPbN** [comment]
Just pushed an additional commit which replaces `***` with `$pwd`.
Rather convenient, as we can then easily copy / past / run the command, after having set `pwd=mypassword` in our current working shell.
Thx for reviewing this PR :)

**Comment by sc979** [comment]
Thanks @CPbN for your work,
I requested some changes on your PR and will launch a job on our CI to test it
Regards
sc979

**Comment by CPbN** [comment]
Many thanks @sc979 for this great review :+1:
I'll push requested changes later today :wink:

File: www/include/monitoring/objectDetails/common-func.php [hunk 1]
```diff
@@ -56,21 +56,28 @@ function hidePasswordInCommand($commandName, $hostId, $serviceId)
 
     // Get command line with macro
     $sth = $pearDB->prepare("SELECT command_line FROM command WHERE command_name = :command_name");
```
[the remark this answers is not in the document]
Comment (CPbN) [reply www/include/monitoring/objectDetails/common-func.php]: I'm not sure `$commandLineWithMacros = ""` is needed.
A few lines below we have `$commandLineWithMacros = $row['command_line']`.
Which makes `$commandLineWithMacros = ""` useless, as `$commandLineWithMacros` will be assigned a string or null value.
And `strlen(null) === 0`, so we're OK.
Feel free if I'm wrong :)

File: www/include/monitoring/objectDetails/common-func.php [hunk 2]
```diff
@@ -91,24 +98,39 @@ function hidePasswordInCommand($commandName, $hostId, $serviceId)
             $query .= ', :svc_id' . $i;
         }
     }
-    $query .= ")";
+    $query .= ") order by svc_macro_name, length(svc_macro_value) desc";
```
[the remark this answers is not in the document]
Comment (CPbN) [reply www/include/monitoring/objectDetails/common-func.php]: We replace macros' values in the command.
Let's assume there are 2 macros in the command, first one with value `MYPASSWORD`, second one with value `MYPASSWORDLONGER`.
In this example, if we begin with the smallest string, we won't then be able to replace the longest one, as this small string is present into the long one, so first macro replacement will also break the second one).
We then need to replace in length desc order.

**Comment by CPbN** [comment]
Don't know why but I forgot to translate messages...
Here they are now.
Thx !

**Comment by CPbN** [comment]
Could you please re-review this when U'll have some time @sc979 ?
Many thanks 👍

**Comment by prohand** [comment]
Hello,

The patch no longer works following the passage in 19.10.4 :

image

Do you have a correction to suggest?

Thank you

**Comment by CPbN** [comment]
Yes this is due to the last commit which patches `.po` files.
Don't bother with them, only patch `www/include/monitoring/objectDetails/common-func.php` and it will be OK !

**Comment by prohand** [comment]
It's OK :

image

Do you have a date when the problem will be finally fixed?

Thank you

**Comment by CPbN** [comment]
I hope to see this in 19.10.5, but then Centreon team needs to finish the review, and merge it 😕

**Comment by prohand** [comment]
Ok thanks ;)

**Comment by CPbN** [comment]
19.10.5 seems ready to go, for 19.10.6 then ? :)

**Comment by cgagnaire** [comment]
Hi @UrBnW,
We discussed about this behaviour on a refinement session.
The dev team will test your work and we will keep you posted ASAP.
Thanks!

**Comment by cgagnaire** [comment]
Refs: MON-4593

**Comment by UrBnW** [comment]
Hi @cgagnaire, really good news, many thanks to you and to the dev team 👍

**Comment by UrBnW** [comment]
Some news perhaps regarding this patch ? Thank you 👍

**Comment by UrBnW** [comment]
Let's finish this PR's review please ?
Thank you !

**Comment by sc979** [comment]
https://github.com/centreon/centreon/pull/8781

**Pull request closed without merging**